### PR TITLE
ARM split - Add serial transport support

### PR DESCRIPTION
**Pull request by zvecr**
## Current Issue
Transport does not exist

## TL;DR
This PR partially ports some of the old lets_split serial code, with slightly different timings.

<details><summary>CLICK ME FOR INVESTIGATION INFO</summary>

## Transport

Current issues include
* drivers/arm/i2c_serial.[ch] not implemented
  * ChibiOS patch has at least been verified against current QMK fork
      * needs verification against latest release
      * self contained master/slave example might help testing
* split_common/serial.[ch] does not compile for ARM

Requirements
* 1 wire protocol
  * Serial
    * complete implementation required
* 2 wire protocol
  * I2C
    * boards may have wired the I2C bus to both a split slave and device like OLED

Upgrading ChibiOS, maintain patch - could break
Potential to lock in to current version if patch cannot be applied going forward

> Think of the current I2C slave code like a patch file, if the code changes the patch might not apply. Or the surrounding code could change behavior, patch applies fine but fails somehow at runtime.

Requires more in-depth knowledge of ChibiOS than "Contributors/Members" have?

### I2C
Pulling the muon code as a intial example, I2C slave support is reasonably self contained. The current implementation manages one block of memory, and reads/writes at offsets. Each offset is handled as a I2C register. The currently exposed API poses an issue due to ChibiOS not providing a mechninism to know the requested number of bytes. When an I2C callback fires, the amount of data to send is unknown.

Potential Solutions
* Stash offsets -> size, sniff register and map to size
* Refactor api to use a structure which can map data with size
    * Each I2C register can then map 
* Refactor to one block, one register
    * Simple
    * No more offsets, transmit/receive all data in one go
    * More overhead
        * chunking based X bytes

### Serial
Currently coded to only support atmega32u4, and contains timings and various settings which are both compiler. Uses AVR specific interrupts and register configuration. Porting it in its current form would be extremely unlikely. Extracting the existing serial.[ch] into a master/slave under `drivers/avr`, with an exposed API. This can then be implemented for ARM, using the relevant ChibiOS calls.

### Alternatives
While outside of current AVR support, anything that is either simple enough to implement, or bundled as part of ChibiOS could be considered. If it can help bootstrap the project, then its on the table.

* RS232/RS485
  * No ChibiOS support
* SPI
  * Untested patch exists on ChibiOS forum

</details>

## Description
Provides a method of transport that does not require a patch or upgrade to ChibiOS. The downside is due to the legacy version of ChibiOS, the code will need to change when we update as interrupts were moved from ext to pal. Stubs have been added for I2C support, to reduce the code changes while #6494 is in progress. Scan rate is reporting as `matrix scan frequency: 692` which is less than ideal but workable for not fast typists (I type ~75wpm and havent seen issues).

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by zvecr** [comment]
## Testing notes

### compatibility
* CTPC works for levinson/rev2
* f103 seems fine on the current changes

### scan rate
32u4 -> 1600
f303 with SERIAL_DELAY 16 -> 779
f303 with SERIAL_DELAY 12 -> 779
f303 with SERIAL_DELAY < 12 -> bad times

**Comment by tzarc** [comment]
Did a test with backlight seeing as I had it set up from the other PRs -- backlight step changes are fine, breathing doesn't transfer over to slave side. Probably needs a note, for now.
